client/internal/httpsender_test.go

```diff
@@ -98,3 +103,73 @@ EKTcWGekdmdDPsHloRNtsiCa697B2O9IFA==
 
 	return cert, nil
 }
+
+func TestHTTPSenderRetryForFailedRequests(t *testing.T) {
+
+	srv, m := newMockServer(t)
+	address := testhelpers.GetAvailableLocalAddress()
+	var connectionAttempts int64
+
+	var buf []byte
+	srv.OnRequest = func(w http.ResponseWriter, r *http.Request) {
+		attempt := atomic.AddInt64(&connectionAttempts, 1)
+		if attempt == 1 {
+			hj, ok := w.(http.Hijacker)
+			if !ok {
+				t.Error("server doesn't support hijacking")
+				return
+			}
+			conn, _, err := hj.Hijack()
+			if err != nil {
+				t.Error(err)
+				return
+			}
+			conn.Close()
+		} else {
+			buf, _ = io.ReadAll(r.Body)
+			w.WriteHeader(http.StatusOK)
+		}
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	url := "http://" + address
+	sender := NewHTTPSender(&sharedinternal.NopLogger{})
+	sender.NextMessage().Update(func(msg *protobufs.AgentToServer) {
+		msg.AgentDescription = &protobufs.AgentDescription{
+			IdentifyingAttributes: []*protobufs.KeyValue{{
+				Key: "service.name",
+				Value: &protobufs.AnyValue{
+					Value: &protobufs.AnyValue_StringValue{StringValue: "test-service"},
+				},
+			}},
+		}
+	})
+	sender.callbacks = types.CallbacksStruct{
+		OnConnectFunc: func() {
+		},
+		OnConnectFailedFunc: func(_ error) {
+		},
+	}
+	sender.url = url
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		sender.sendRequestWithRetries(ctx)
+		wg.Done()
+	}()
+	go func() {
+		time.Sleep(200 * time.Millisecond)
```

Comment:
I'm concerned this won't be completely reliable. My understanding is that this is intended to wait for the request to fail before starting a server, is there a way we can wait for the `OnConnectFailed` callback to fire before starting it?

Reply:
+1

Reply:
I added this when trying to create a failed test based on this discussion https://github.com/open-telemetry/opamp-go/pull/194#discussion_r1300231060. This is no longer needed because the connection refused error doesn't reproduce the original issue so I updated the test to create a connection reset issue. Hope this is clear.